src/sql/pool.rs

```diff
@@ -0,0 +1,105 @@
+//! Connection pool.
+
+use std::fmt;
+use std::ops::{Deref, DerefMut};
+use std::sync::{Arc, Weak};
+
+use parking_lot::{Condvar, Mutex};
+use rusqlite::Connection;
+
+/// Inner connection pool.
+struct InnerPool {
+    /// Available connections.
+    connections: Mutex<Vec<Connection>>,
```

Comment:
If the intent is to have a fixed number of connections at compile time, why not make this an array and the pool generic over `const SIZE: usize` and move the init of the connection into the constructor?

Reply:
this is how the array version would look like: https://github.com/deltachat/deltachat-core-rust/commit/c2e275cb967458448e37d778a4e0cbbc032d7d9e

If needed the mutex on all connections could be removed, and replaced with an atomicptr construction on each element, which would allow concurrent retrieval of connections, as there is no "global" access on the array needed anymore, given its fixed size.